Frozen lockfile setting in CI can't be overridden via env var or updateConfig hook

### Verify latest release

- [x] I verified that the issue exists in the latest pnpm release

### pnpm version

_No response_

### Which area(s) of pnpm are affected? (leave empty if unsure)

_No response_

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/TrevorBurnham/pnpm-frozen-lockfile-bug-repro

### Reproduction steps

In [the linked repository](https://github.com/TrevorBurnham/pnpm-frozen-lockfile-bug-repro), run this command:

```
CI=true pnpm install
```

### Describe the Bug

When the env var `CI` is `true`, `pnpm install` throws an error if the `pnpmfileChecksum` in `pnpm-lock.yaml` doesn't match the contents of the pnpmfile:

> ERR_PNPM_LOCKFILE_CONFIG_MISMATCH  Cannot proceed with the frozen installation. The current "pnpmfileChecksum" configuration doesn't match the value found in the lockfile

That's perfectly consistent with the [documented behavior of `--frozen-lockfile`](https://pnpm.io/cli/install#--frozen-lockfile). However, what's strange is how hard to override this behavior is:

* ❌ `CI=true npm_config_frozen_lockfile=false pnpm install` fails
* ❌ `CI=true pnpm_config_frozen_lockfile=false pnpm install` fails
* ❌ Using the `updateConfig` hook in `pnpmfile.js` to set `frozenLockfile: false` fails.
* ✅ Creating a `.npmrc` with `frozen-lockfile = false` works.
* ✅ `CI=true pnpm install --frozen-lockfile=false` works.
* ✅ `CI=true pnpm install --no-frozen-lockfile` works.

### Expected Behavior

I'd expect setting `frozenLockfile` via an env var or the `updateConfig` hook to have the same effect as setting it via `.npmrc` or a CLI flag.

### Which Node.js version are you using?

v22.17.1

### Which operating systems have you used?

- [x] macOS
- [ ] Windows
- [ ] Linux

### If your OS is a Linux based, which one it is? (Include the version if relevant)

_No response_

--- a/pkg-manager/plugin-commands-installation/src/install.ts
+++ b/pkg-manager/plugin-commands-installation/src/install.ts
@@ -361,8 +361,8 @@ export async function handler (opts: InstallCommandOptions): Promise<void> {
     ...opts,
     frozenLockfileIfExists: opts.frozenLockfileIfExists ?? (
       opts.ci && !opts.lockfileOnly &&
-      typeof opts.rawLocalConfig['frozen-lockfile'] === 'undefined' &&
-      typeof opts.rawLocalConfig['prefer-frozen-lockfile'] === 'undefined'
+      typeof opts.frozenLockfile === 'undefined' &&
+      typeof opts.preferFrozenLockfile === 'undefined'
     ),
     include,
     includeDirect: include,

--- a/pnpm/test/install/misc.ts
+++ b/pnpm/test/install/misc.ts
@@ -437,6 +437,58 @@ test('installing in a CI environment', async () => {
   await execPnpm(['install', '--no-prefer-frozen-lockfile'], { env: { CI: 'true' } })
 })
 
+// Tests for issue #9861: frozen-lockfile should be overridable via env vars and updateConfig hook
+test('CI mode: frozen-lockfile can be overridden via environment variable', async () => {
+  const project = prepare({
+    dependencies: { rimraf: '2.5.1' },
+  })
+
+  // Initial install in CI mode
+  await execPnpm(['install'], { env: { CI: 'true' } })
+
+  // Change dependencies
+  project.writePackageJson({
+    dependencies: { rimraf: '1' },
+  })
+
+  // Should not fail when pnpm_config_frozen_lockfile is set to false
+  await execPnpm(['install'], {
+    env: {
+      CI: 'true',
+      pnpm_config_frozen_lockfile: 'false',
+    },
+  })
+})
+
+test('CI mode: frozen-lockfile can be overridden via updateConfig hook', async () => {
+  const project = prepare({
+    dependencies: { rimraf: '2.5.1' },
+  })
+
+  const pnpmfile = `
+    module.exports = {
+      hooks: {
+        updateConfig(config) {
+          config.frozenLockfile = false
+          return config
+        }
+      }
+    }
+  `
+  fs.writeFileSync('.pnpmfile.cjs', pnpmfile, 'utf8')
+
+  // Initial install in CI mode
+  await execPnpm(['install'], { env: { CI: 'true' } })
+
+  // Change dependencies
+  project.writePackageJson({
+    dependencies: { rimraf: '1' },
+  })
+
+  // Should not fail due to updateConfig hook setting frozenLockfile to false
+  await execPnpm(['install'], { env: { CI: 'true' } })
+})
+
 test('installation fails with a timeout error', async () => {
   prepare()
 

